### ISSUE TITLE
Does "weave reset" really forget everything?

### ISSUE THREAD

**Issue by igenyar** [weaveworks/weave#1292]
I am doing some experiment to learn the usage of weave.

One experiment does simple TCP socket client/server send/recv, like below (client side).

            while (1)
            { n = send(sock, "hello", 5, 0);
              printf("client >    n= %d\n", n);
              getchar();
            }

Basically, I let socket open and keep sending strings but wait for user input to continue. During the meantime, I insert "weave reset/launch/expose" a few times at the client side.

What I see is: (a) that if I keep weave IP unchanged, the socket seems not broken (sending/receiving good). But (b) if I modify weave IP, there is no error and it seems it can still send even though the server does not receive anything. More amazingly: (c) that if I change weave IP back again, the server receives everything in between as if there were a buffer.

In short, "weave reset" does not forget everything and in some cases the socket should be broken and it should report error.

BTW,

    $ weave version
    weave script 1.0.1
    weave router 1.0.1
    Unable to find weaveworks/weavedns:1.0.1 image.
    weave exec 1.0.1

**Comment by rade** [comment]
Please provide us with the exact steps to reproduce this.

`weave reset` removes the weave bridge. This in turn removes the `ethwe` interface from the containers. And it won't come back with `weave launch`.

````
$ weave launch
0bac0175fba648c46e27df640090cc74ebb009c948d7ac58f1edb4d534f91b3f
$ weave run --name=foo -ti ubuntu
50f62765a6cdf90029737187f0ffeddac686430a899f5451f3d0266899b3520f
$ docker exec foo ifconfig ethwe
ethwe     Link encap:Ethernet  HWaddr e6:75:a2:48:45:13  
          inet addr:10.128.0.1  Bcast:0.0.0.0  Mask:255.192.0.0
          inet6 addr: fe80::e475:a2ff:fe48:4513/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:65535  Metric:1
          RX packets:23 errors:0 dropped:0 overruns:0 frame:0
          TX packets:9 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:3758 (3.7 KB)  TX bytes:690 (690.0 B)

$ weave reset
$ docker exec foo ifconfig ethwe
ethwe: error fetching interface information: Device not found
$ weave launch
010c3c0a196d43c3b23af670893b5439390d50f153e5a2bdc2e5bd34935e3605
$ docker exec foo ifconfig ethwe
ethwe: error fetching interface information: Device not found
````

**Comment by rade** [comment]
You mention "keeping" or "modifying" the weave IP. Which suggests that you are perhaps `weave attach`ing the container after re-launching weave. TCP/IP is very resilient to network outages, so it wouldn't surprise me if connections survived the temporary disappearance of network interfaces.

**Comment by igenyar** [comment]
The server code looks like below. For brevity, I skip the part of socket, bind, listen, accept... nothing really fancy there.

                    while (1)
                    { int n = recv(_sock, buffer, 1024, 0);
                      if (n > 0)
                      { buffer[n] = '\0';
                        printf("received: %s\n", buffer);
                      }
                    }

Once the connection is established after accept, the server will constantly recv.

I use three machines: A, B and C.

On C,
    $ weave launch

On A,
    $ weave reset
    $ weave launch C_addr
    $ weave expose 10.2.0.101/16
    Then run server code which listens at a port. Keep the server running.

On B
    $ weave reset
    $ weave launch C_addr
    $ weave expose 10.2.0.102/16

Open another terminal (on B), and run client code.
The client will connect server via server's weave IP. Keep the client running.
Note, I can control client send by "getchar()".

    1st send/receive -- good
    do weave reset, use the same IP
        $ weave reset
        $ weave launch C_addr
        $ weave expose 10.2.0.102/16
    2nd send/receive -- still good, observation (a) in my original post
    do weave reset, use a different IP
        $ weave reset
        $ weave launch C_addr
        $ weave expose 10.2.0.103/16
    3rd send/receive -- send returns good, but server not receiving, observation (b)
    4th send/receive -- send returns good, but server not receiving, observation (b)
    5th send/receive -- send returns good, but server not receiving, observation (b)
    do weave reset, change the address back
        $ weave reset
        $ weave launch C_addr
        $ weave expose 10.2.0.102/16
    6th send/receive -- send/recv good, even including previous ones! observation (c)

The output at the server side looks like this.

    received: hello
    received: hello
    received: hellohellohellohello

**Comment by igenyar** [comment]
No, "weave reset" removes container, how to "re-attach" something which does not exist anymore?

**Comment by rade** [comment]
I see you are not actually running the send/recv code in containers, and are instead using `weave expose` and run it on the host. What I said about TCP/IP resilience equally applies here.

> No, "weave reset" removes container, how to "re-attach" something which does not exist anymore?

I had assumed you were running the send/recv code in containers. `weave reset` does *not* remove application containers.
